### ISSUE TITLE
WAL data missing in backup restoration

### ISSUE THREAD

**Issue by Tekchanddagar** [pgbackrest/pgbackrest#1616]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version: 2.37

2. PostgreSQL version: 12

3. Operating system/version - Ubuntu 20.04

4. Did you install pgBackRest from source or from a package? Package

5. Describe the issue:

We are using PgBackrest version 2.37 and using S3 repository as backup storage for full backup, differential backup as well as WAL archive.

We are using `archive-async` method for WAL archive. Below is the our `pgbackrest.conf` config:

```
[global]
repo1-type=s3
repo1-s3-bucket=database-pgbackrest
repo1-path=/eva-db
repo1-s3-endpoint=s3.eu-central-1.amazonaws.com
repo1-s3-key=************************
repo1-s3-key-secret=*************************************
repo1-s3-region=eu-central-1
repo1-retention-full=5
repo1-retention-diff=3
compress-type=zst
start-fast=y
compress-level-network=3
# used where not overridden
process-max=4
archive-push-queue-max=5GB
archive-async=y
spool-path=/var/spool/pgbackrest

[eva]
pg1-path=/var/lib/postgresql/12/main
log-level-console=error
log-level-file=info

[global:archive-push]
compress-level=3
# more cores for archive-push
process-max=4

[global:archive-get]
# fewer cores for archive-get
process-max=8

[global:backup]
# more cores for backup
process-max=4

[global:restore]
# all the cores for restore
process-max=8

``` 
Below are the latest log for `archive-push-async.log`. From logs its seems that WAL are archive in S3 bucket and we can see the same WAL in S3 bucket as well. The time in log is `UTC`
```
------------------PROCESS START-------------------
2022-01-06 17:59:15.947 P00   INFO: archive-push:async command begin 2.37: [/var/lib/postgresql/12/main/pg_wal] --archive-async --archive-push-queue-max=5GB --compress-level=3 --compress-level-network=3 --compress-type=zst --exec-id=18967-5a2df7eb --log-level-console=off --log-level-file=info --log-level-stderr=off --pg1-path=/var/lib/postgresql/12/main --process-max=4 --repo1-path=/eva-db --repo1-s3-bucket=database-pgbackrest --repo1-s3-endpoint=s3.eu-central-1.amazonaws.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=eu-central-1 --repo1-type=s3 --spool-path=/var/spool/pgbackrest --stanza=eva
2022-01-06 17:59:15.948 P00   INFO: push 1 WAL file(s) to archive: 0000000A0000042500000018
2022-01-06 17:59:16.672 P00   INFO: archive-push:async command end: completed successfully (726ms)

```

Now when we are restoring the backup on new machine its stored successfully and when we are starting the postgresql service its also running fine.

But when we are checking the data on new machine where we restored the backup we found that lot of data is not there. But we are expecting only few minutes data might be not there. 

As per our understanding during restore it will use `Full Backup+Diffrential Backup+WAL`. So only few WAL data should be not there.

In the `archive-push-async.log` we can see the time is `2022-01-06 17:59:16.672` but when we are checking the data on new machine we are found the data of till `2022-01-06 6:36:56.672` only. We took the full backup at `6 AM UTC`:

```
root@ubuntu-db:~# sudo -u postgres pgbackrest --stanza=eva --log-level-console=info info
stanza: eva
    status: ok
    cipher: none

    db (current)
        wal archive min/max (12): 0000000A0000042100000053/0000001100000421000000A6

        full backup: 20220106-060055F
            timestamp start/stop: 2022-01-06 06:00:55 / 2022-01-06 06:06:41
            wal start/stop: 0000000A0000042100000059 / 0000000A0000042100000066
            database size: 94.3GB, database backup size: 94.3GB
            repo1: backup set size: 13.5GB, backup size: 13.5GB

```
We can see the data till `2022-01-06 6:36:56.672` on new machine so some WAL are also restored, but we are expecting all the WAL should be restored.

Are we missing any parameter in our config?  Can you please help us?

Thank You

**Comment by dwsteele** [comment]
What was the complete restore command that you ran? Also, you didn't include PostreSQL and pgBackRest logs.

To be clear, pgBackRest does *not* do recovery, PostgreSQL does. In our experience the vast majority of recovery "issues" are because the user has not understood the pgBackRest and/or PostgreSQL documentation.

So, we'll help if we can, but the issue is very unlikely to be related to pgBackRest.

**Comment by Tekchanddagar** [comment]
@dwsteele, Thank You for your response.


> What was the complete restore command that you ran?

First we clean the data directory of server on which we are restoring the data. Then we ran the below command to restore the backup:
```
sudo -u postgres pgbackrest --stanza=eva restore --log-level-console=detail
```
Logs for `pgbackrest`

```
2022-01-10 04:32:49.624 P00   INFO: repo1: restore backup set 20220107-045529F_20220107-121833D, recovery will start at 2022-01-07 12:18:33

2022-01-10 04:34:10.571 P00   INFO: restore global/pg_control (performed last to ensure aborted restores cannot be started)
2022-01-10 04:34:10.572 P00 DETAIL: sync path '/var/lib/postgresql/12/main/global'
2022-01-10 04:34:10.572 P00   INFO: restore size = 95.3GB, file total = 1993
2022-01-10 04:34:10.573 P00 DETAIL: statistics: {"http.client":{"total":1},"http.request":{"total":2},"http.session":{"total":1},"socket.client":{"total":1},"socket.session":{"total":1},"tls.client":{"total":1},"tls.session":{"total":1}}
2022-01-10 04:34:10.573 P00   INFO: restore command end: completed successfully (81140ms)
```

Logs for `postgresql`
```
Jan 10 04:34:46 eva-db-03 postgres[66583]: [1-1] 2022-01-10 04:34:46.126 UTC [66583] LOG:  starting PostgreSQL 12.9 (Ubuntu 12.9-2.pgdg20.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0, 64-bit
Jan 10 04:34:46 eva-db-03 postgres[66583]: [2-1] 2022-01-10 04:34:46.126 UTC [66583] LOG:  listening on IPv4 address "10.133.253.18", port 5432
Jan 10 04:34:46 eva-db-03 postgres[66583]: [3-1] 2022-01-10 04:34:46.128 UTC [66583] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
Jan 10 04:34:46 eva-db-03 postgres[66583]: [4-1] 2022-01-10 04:34:46.185 UTC [66583] LOG:  ending log output to stderr
Jan 10 04:34:46 eva-db-03 postgres[66583]: [4-2] 2022-01-10 04:34:46.185 UTC [66583] HINT:  Future log output will go to log destination "syslog".
Jan 10 04:34:46 eva-db-03 postgres[66584]: [5-1] 2022-01-10 04:34:46.189 UTC [66584] LOG:  database system was interrupted; last known up at 2022-01-07 12:18:34 UTC
Jan 10 04:34:46 eva-db-03 postgres[66584]: [6-1] 2022-01-10 04:34:46.367 UTC [66584] LOG:  restored log file "0000000B.history" from archive
Jan 10 04:34:46 eva-db-03 postgres[66584]: [7-1] 2022-01-10 04:34:46.457 UTC [66584] LOG:  starting archive recovery
Jan 10 04:34:46 eva-db-03 postgres[66584]: [8-1] 2022-01-10 04:34:46.556 UTC [66584] LOG:  restored log file "0000000B.history" from archive
Jan 10 04:34:48 eva-db-03 postgres[66584]: [9-1] 2022-01-10 04:34:48.687 UTC [66584] LOG:  restored log file "0000000A00000428000000D2" from archive
Jan 10 04:34:48 eva-db-03 postgres[66584]: [10-1] 2022-01-10 04:34:48.703 UTC [66584] FATAL:  requested timeline 11 is not a child of this server's history
Jan 10 04:34:48 eva-db-03 postgres[66584]: [10-2] 2022-01-10 04:34:48.703 UTC [66584] DETAIL:  Latest checkpoint is at 428/D2000060 on timeline 10, but in the history of the requested timeline, the server forked off from that timeline at 428/C1FFEC88.
Jan 10 04:34:48 eva-db-03 postgres[66583]: [5-1] 2022-01-10 04:34:48.706 UTC [66583] LOG:  startup process (PID 66584) exited with exit code 1
Jan 10 04:34:48 eva-db-03 postgres[66583]: [6-1] 2022-01-10 04:34:48.706 UTC [66583] LOG:  aborting startup due to startup process failure
Jan 10 04:34:48 eva-db-03 postgres[66583]: [7-1] 2022-01-10 04:34:48.715 UTC [66583] LOG:  database system is shut down
```
We have checked `archive-get-async.log` but didn't found any error.

We have ran the `verify` command also and the output shows no wal is missing:

```
va-db-01:/var/spool/pgbackrest/archive/eva/out# sudo -u postgres pgbackrest --stanza=eva --log-level-console=info verify
2022-01-10 04:56:00.378 P00   INFO: verify command begin 2.37: --compress-level-network=3 --exec-id=27901-5216a4ed --log-level-console=info --log-level-file=info --process-max=4 --repo1-path=/eva-db --repo1-s3-bucket=sf-database-pgbackrest --repo1-s3-endpoint=s3.eu-central-1.amazonaws.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=eu-central-1 --repo1-type=s3 --stanza=eva
2022-01-10 05:03:10.216 P00   INFO: Results:
                                      archiveId: 12-1, total WAL checked: 3540, total valid WAL: 3540
                                        missing: 0, checksum invalid: 0, size invalid: 0, other: 0
                                      backup: 20220107-045529F, status: valid, total files checked: 1992, total valid files: 1992
                                        missing: 0, checksum invalid: 0, size invalid: 0, other: 0
                                      backup: 20220107-045529F_20220107-121833D, status: valid, total files checked: 1993, total valid files: 1993
                                        missing: 0, checksum invalid: 0, size invalid: 0, other: 0
2022-01-10 05:03:10.216 P00   INFO: verify command end: completed successfully (429839ms)
```

But when we are checking the `wal stats` using the `pg_stat_archiver` its show `12` WAL are missing. And we are aware that if single `WAL` is missing in S3 repo then restoration will be fail.

```
archived_count |    last_archived_wal     |      last_archived_time       | failed_count |     last_failed_wal      |       last_failed_time        |          stats_reset          
----------------+--------------------------+-------------------------------+--------------+--------------------------+-------------------------------+-------------------------------
           5567 | 0000000A000004340000006A | 2022-01-10 05:30:49.212077+00 |           12 | 0000000A0000042600000054 | 2022-01-07 04:54:04.374943+00 | 2022-01-05 13:37:07.688426+00

```

The strange thing is that when we performed the restoration after only full backup then restoration was successfull. Then we took differential backup on 7th Jan:

```
eva-db-01:/var/log/pgbackrest# sudo -u postgres pgbackrest --stanza=eva info --log-level-console=info
stanza: eva
    status: ok
    cipher: none

    db (current)
        wal archive min/max (12): 0000000A0000042600000054/0000000B00000428000000C2

        full backup: 20220107-045529F
            timestamp start/stop: 2022-01-07 04:55:29 / 2022-01-07 05:00:10
            wal start/stop: 0000000A0000042600000058 / 0000000A000004260000005A
            database size: 95.0GB, database backup size: 95.0GB
            repo1: backup set size: 13.6GB, backup size: 13.6GB

        diff backup: 20220107-045529F_20220107-121833D
            timestamp start/stop: 2022-01-07 12:18:33 / 2022-01-07 12:22:14
            wal start/stop: 0000000A00000428000000D2 / 0000000A00000428000000D5
            database size: 95.3GB, database backup size: 68.2GB
            repo1: backup set size: 13.7GB, backup size: 10.8GB
            backup reference list: 20220107-045529F
```

And now we are performing restoration again and it is failing again.

The strange thing is that the output of `verify` and `pg_stat_archiver` its show` is different and `pg_stat_archiver` its show` command is showing some archived failed. At same time we are not aware how `verify` and `pg_stat_archiver` commands works.

Can you please help us now? Please let us know if you need any other info from our side.

Thank You

**Comment by pgstef** [comment]
Hi,

For your first restore, given this info:
```
wal archive min/max (12): 0000000A0000042100000053/0000001100000421000000A6

full backup: 20220106-060055F
    timestamp start/stop: 2022-01-06 06:00:55 / 2022-01-06 06:06:41
    wal start/stop: 0000000A0000042100000059 / 0000000A0000042100000066
    database size: 94.3GB, database backup size: 94.3GB
    repo1: backup set size: 13.5GB, backup size: 13.5GB
```

The PG error is pretty clear:
```
FATAL:  requested timeline 11 is not a child of this server's history
DETAIL:  Latest checkpoint is at 428/D2000060 on timeline 10, but in the history of the requested timeline, 
the server forked off from that timeline at 428/C1FFEC88.
```

Basically, there's a timeline mismatch. Your backup is on `0000000A` (10 in hexadecimal) and you request to follow 11 (because you somehow have `00000011.history` + 00000011... wals). But the timeline switch point seems to be before your full backup. Not sure how you got that situation though. Probably a test restore generated that new timeline and you took a new full backup afterwards.
It's often recommended to use the [--archive-mode=off](https://pgbackrest.org/configuration.html#section-restore/option-archive-mode) option for a test restore, which will disable the archiving system on the restored cluster to avoid messing with the actual repo.

> When we performed the restoration after only full backup then restoration was successfull and `WAL REPLAY` was also working

```
LOG:  restored log file "0000000A00000428000000C1" from archive
LOG:  selected new timeline ID: 11
LOG:  archive recovery complete
```

So, that's where you got your timeline 11.

Then you took another backup which is also on timeline `0000000A`:
```
diff backup: 20220107-045529F_20220107-121833D
    timestamp start/stop: 2022-01-07 12:18:33 / 2022-01-07 12:22:14
    wal start/stop: 0000000A00000428000000D2 / 0000000A00000428000000D5
    database size: 95.3GB, database backup size: 68.2GB
    repo1: backup set size: 13.7GB, backup size: 10.8GB
    backup reference list: 20220107-045529F
```

And you're probably getting the same `requested timeline 11 is not a child of this server's history` error. That's not stating there are missing wals... Not sure why you think this.

Basically speaking, your "production" database and your repo content don't match. PG recovery will see the `00000011.history` and since the PG `recovery_target_timeline` setting is **latest**, it will try to follow that timeline. The timeline switch happened before your backup which is on timeline 10. So for sure, that won't work. You have to clean the timeline 11 from your repo (and probably every other timeline generated afterwards) or try `--target-timeline=current`.

Finally, the `verify` command is still ["internal"](https://github.com/pgbackrest/pgbackrest/blob/main/src/build/config/config.yaml#L154) and not described on the [docs website](https://pgbackrest.org/command.html) because it's not complete. It's actually verifying the checksums but, afaik, not if there are missing archives. (Side note, [check_pgbackrest](https://github.com/pgstef/check_pgbackrest) could probably tell you)

`pg_stat_archiver` is only giving you the amount of failed archive_command. And when the archive_command fails, PG tries it again until it succeeds (keeping the failing wal in its pg_wal directory until it has been archived). So, once again, no clue telling you there are missing wals.

Regards,

**Comment by Tekchanddagar** [comment]
@dwsteele, Thank You for your help.
You guide us in right direction, we enabled the `archive` on test restore machine due to timeline got changed when we start the postgresql service on test restore machine then it change the data on S3 as well.
 
We have tried both the option "recovery_target_timeline setting is latest" and by disabling the `--archive-mode=off` and both are working fine.

Take a bow :)

Thank You
